### PR TITLE
Process isolation, consistent log levels, and avoiding groovy version issues

### PR DESCRIPTION
**Pull request by lukebemish**
Among other changes, the conversion task now runs a "runner" layer which runs different conversion tasks on threads, and is started up once in another process as a build service. This system means that we only fork the JVM once instead of for every mods.groovy generation task. Additionally, types needed by the gradle plugin have been moved to the new "types" artifact, to avoid pulling in a bad groovy version on the plugin classpath. Additionally, logging is fixed to obey externally set log levels and the gradle log level is propogated in for the plugin.

### CODE REVIEW

File: types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        return clazz.isArray() ? allows(clazz.componentType()) : ALLOWED_CLASSES.contains(clazz);
+    }
+
+    private static final Set<Class<?>> ALLOWED_CLASSES = Set.of(
+            String.class,
+
+            // Collections
+            LinkedHashMap.class,
+            HashMap.class,
+            ArrayList.class,
```
Comment (Matyrobbrt) [inline types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java:49]: I would add (linked)hashsets here too.

File: types/build.gradle [hunk 2]
```diff
@@ -0,0 +1,19 @@
+dependencies {
+    compileOnly libs.jetbrains.annotations.get()
```
Comment (Matyrobbrt) [inline types/build.gradle:2]:
```suggestion
    compileOnly libs.jetbrains.annotations
```

File: types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java
Comment (Matyrobbrt) [inline types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java:1]: Why was this moved?

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: types/build.gradle (hunk 2 above)
Comment (PaintNinja) [reply types/build.gradle:2]: Last I checked, while the `.get()` isn't necessary for it to work with Gradle, IntelliJ shows an incompatible types warning when the get is missing

File: types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java (hunk 1 above)
Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java:49]: We don't pass any, and the data is sanitized before it gets passed to contain only the stuff in question. I did realize that I don't do this sanitization on the runner -> gradle end -- I should do that

File: types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java
Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java:1]: Because it's used in MDGExtension

File: types/build.gradle (hunk 2 above)
Comment (lukebemish) [reply types/build.gradle:2]: On newer intellij versions, this seems to either work fine for `.get()` or no `.get()`, _or_ (as seems to be the case for this one) gets no IDE support -- maybe because the version catalog is in code, or maybe because of the includeBuild? It's unclear

File: types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java (hunk 1 above)
Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java:49]: Data is now sanitized on both ends.

File: types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java
Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java:1]: I would just depend on flexver outright but it's not on maven central... not sure how to best solve this. It's a small file -- could just have it in both places?

Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java:1]: Shoving this in an internal package and marking it as such -- I think that's really the simplest solution here, because plugin and core both need types anyways so it might as well be there.

Comment (PaintNinja) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java:1]: If we're moving stuff around, I'd like us to consider another PR in the future for JPMS support - at least not having the same package spread out across multiple things - a mistake I made during early design of v2 when I was unaware of that module restriction. Separate PR to make it easier to keep track of and not make this one even larger.

Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/core/versioning/FlexVerComparator.java:1]: The only incompliant bit in that regard is the plugins -- everything else should already be in its own packages, except the frontends which can't be because they have stuff in the root package.

File: runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            }
+        })
+    }
+
+    @TupleConstructor(includeFields = true)
+    private static class Output {
+        private final ObjectOutputStream stream
+
+        synchronized void writeObject(Object obj) {
+            stream.writeObject(obj)
```
Comment (PaintNinja) [inline runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy]: Could this be a private record instead?

File: runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                } else {
+                    throw new IOException("Unexpected object: " + obj)
+                }
+            } catch (ClassNotFoundException e) {
+                throw new IOException(e)
+            }
+        }
+    }
+
+    private static final CompilerConfiguration MDG_COMPILER_CONFIG = new CompilerConfiguration().tap {
```
Comment (PaintNinja) [inline runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy]: Please keep static final fields at the top of classes so that they're easy to find

File: types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            String.class,
+            Date.class,
+
+            // Collections
+            LinkedHashMap.class,
+            HashMap.class,
+            ArrayList.class,
+
+            // Primitives
+            Integer.class,
```
Comment (PaintNinja) [inline types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java:52]: Should we also include `int.class`, `float.class`, etc?

File: runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy (hunk 3 above)
Comment (lukebemish) [reply runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy]: Yes-ish. Technically, yes, but on the other end marking an ObkectOutputStream as immutable (which is what you'd have to do for a record -- groovy records require each included type to be marked immutable) seems wrong

File: types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java (hunk 5 above)
Comment (lukebemish) [reply types/src/main/java/org/groovymc/modsdotgroovy/types/runner/FilteredStream.java:52]: No. Those are never looked up

File: runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy (hunk 3 above)
Comment (PaintNinja) [reply runner/src/main/groovy/org/groovymc/modsdotgroovy/runner/ModsDotGroovyRunner.groovy]: Ah, I forgot about that :/
Could we at least add the `final` modifier to the class?
